### DISCUSSION TITLE
Serverless Compose gives us "No integration defined for method" error

### DISCUSSION THREAD

**Discussion by bronzeson** [serverless/serverless#11136] [Q&A]
Just wondering if anyone else has seen this while using Serverless Compose.
It could be that it's our specific setup, but if others are seeing it too then it might be a problem with Compose.

We recently switched to using serverless compose for deploying our services.
We use a monorepo for multiple services. They all share the same API Gateway. We see the following error seemingly randomly:

`InteractionService › CREATE_FAILED: ApiGatewayDeployment1654529537280 (AWS::ApiGateway::Deployment)
InteractionService › Resource handler returned message: "No integration defined for method (Service: ApiGateway, Status Code: 400, Request ID: xxxxx)" (RequestToken: xxxxx, HandlerErrorCode: InvalidRequest)`

We theorize that something is going on where things try to get hooked up to the lambdas before they are made and are wondering if we are overloading ApiGateway or if something gets messed up when using Compose

**Comment by mnapoli** [comment]
> They all share the same API Gateway.

Just to be sure, do you create the shared API Gateway first, and then (via `dependsOn`) deploy the other services?

> are wondering if we are overloading ApiGateway

Interesting, do you know how many services are deploying in parallel (services that would impact the same API Gateway in parallel)?

cc @pgrzesik

**Reply by bronzeson** [reply]
We create it manually beforehand (what we actually do is use serverless to deploy and create it, then grab the IDs we need to reference it, but we can for the sake of this discussion consider it to be created beforehand manually).

We deploy 2 layers and 8 services with compose. All the services depend on both layers and nothing else. 7 of the services depend on API Gateway

**Reply by pgrzesik** [reply]
Hey @bronzeson - thanks for reporting and sorry you've run into trouble. This seems to be a known problem with parallel deployments of multiple services that share the same API Gateway - please see here as well - https://github.com/serverless/serverless/issues/9036

We don't have a solution for it and so far the only workaround is to deploy services in sequence or in smaller parallel batches to avoid these errors from APIGW. Another discussed approach is to not use shared API Gateway in such settings but that probably isn't an option for you at

**Reply by bronzeson** [reply]
Thanks for the response!
Do y'all have any plans to try to solve the issue in the future? Or is it purely an ApiGateway limitation? If it is a limitation that you don't plan on solving soon, we'll probably prioritize doing something like this: https://seed.run/blog/how-to-share-an-api-domain-between-services-in-a-monorepo-serverless-app.html

**Reply by mnapoli** [reply]
@bronzeson this isn't a direct solution but you can now use `--max-concurrency` to limit the number of services deployed in parallel: https://github.com/serverless/compose/issues/134

That may help avoid the error here?

**Reply by pgrzesik** [reply]
We don't have the definitive answer, but it seems like it might be a limitation of APIGW - all our updates go through CloudFormation and if they happen in parallel, it seems like they might interfere with each other. If you have access to AWS support, you might want to ask them about this behavior and they should be able to provide more guidance.

**Reply by Breadkenty** [reply]
Anybody have any updates on this? We have a monorepo solution as well as feature branch builds that we deploy as a stage (prod, staging, dev, and feature-branch) that require us to use one api gateway instance. I've set concurrency to 4 and it randomly throws this error. The worst part is when a deployment fails that way, the stack results in the status `ROLLBACK_COMPLETE` which forces my CI pipeline code to `serverless remove` before running `serverless deploy` again.

**Reply by pgrzesik** [reply]
Hey @Breadkenty. This is unfortunately something that we have very little control over, as this is a problem with AWS API Gateway that is being updated concurrently, it sometimes throws this error and it has been confirmed by multiple people (also by AWS support). If you want to reduce it happening, you need to reduce the number of concurrent updates to APIGW resources.

**Reply by ShlomoVinny** [reply]
>  you need to reduce the number of concurrent updates to APIGW resources

So theoretically, if you use --max-concurrency 1  it shouldnt throw errors at all, since youre only updating APIGW with 1 resource at a time?

**Reply by pgrzesik** [reply]
Hey @ShlomiVinny 👋 Yes, that should be the case. Of course, that limits your ability to update resources concurrently. You can also define explicit dependencies between services that use APIGW in a way that will force them to deploy sequentially
